### PR TITLE
TT-11285 implement dns refresh mechanism

### PR DESCRIPTION
**Pull request by sredxny**
### **User description**
<details open>
  <summary><a href="https://tyktech.atlassian.net/browse/TT-11285" title="TT-11285" target="_blank">TT-11285</a></summary>
  <br />
  <table>
    <tr>
      <th>Summary</th>
      <td>Hybrid Tyk Gateway fails to reconnect RPC when remote IP becomes unavailable</td>
    </tr>
    <tr>
      <th>Type</th>
      <td>
        Bug
        Bug
      </td>
    </tr>
    <tr>
      <th>Status</th>
      <td>In Code Review</td>
    </tr>
    <tr>
      <th>Points</th>
      <td>N/A</td>
    </tr>
    <tr>
      <th>Labels</th>
      <td><a href="https://tyktech.atlassian.net/issues?jql=project%20%3D%20TT%20AND%20labels%20%3D%20DoD_Fail%20ORDER%20BY%20created%20DESC" title="DoD_Fail">DoD_Fail</a>, <a href="https://tyktech.atlassian.net/issues?jql=project%20%3D%20TT%20AND%20labels%20%3D%20jira_escalated%20ORDER%20BY%20created%20DESC" title="jira_escalated">jira_escalated</a>, <a href="https://tyktech.atlassian.net/issues?jql=project%20%3D%20TT%20AND%20labels%20%3D%20rel5-2025-candidate-commercial%20ORDER%20BY%20created%20DESC" title="rel5-2025-candidate-commercial">rel5-2025-candidate-commercial</a>, <a href="https://tyktech.atlassian.net/issues?jql=project%20%3D%20TT%20AND%20labels%20%3D%20security_skip%20ORDER%20BY%20created%20DESC" title="security_skip">security_skip</a></td>
    </tr>
  </table>
</details>

---

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Refactoring or add test (improvements in base code or adds test coverage to functionality)

## Checklist

- [ ] I ensured that the documentation is up to date
- [ ] I explained why this PR updates go.mod in detail with reasoning why it's required
- [ ] I would like a code coverage CI quality gate exception and have explained why

___

### **PR Type**
Bug fix, Enhancement, Tests

___

### **Description**
- Implements DNS resolution refresh and self-healing for RPC connections

- Adds logic to detect DNS changes and trigger client reconnection

- Introduces robust error handling for network-related RPC failures

- Provides comprehensive unit tests for DNS and utility logic

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>dns_resolver.go</strong><dd><code>Add DNS resolution and reconnection logic for RPC</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

rpc/dns_resolver.go

<li>Introduces DNSResolver interface and default implementation<br> <li> Adds logic to detect DNS changes and trigger RPC client reconnection<br> <li> Provides thread-safe tracking of last resolved IPs<br> <li> Exposes functions for DNS checking and reconnection logic

</details>

  </td>
  <td><a href="https://github.com/TykTechnologies/tyk/pull/7136/files#diff-3d4fcf2e5b1a8e522dcc0dc656f64dfa85e53f41e8c46c5466d422b8d56d8a82">+130/-0</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>rpc_client.go</strong><dd><code>Integrate DNS refresh and error handling in RPC client</code>&nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

rpc/rpc_client.go

<li>Tracks DNS check state after connection errors<br> <li> Integrates DNS refresh on connection and after login<br> <li> Adds error handler to trigger DNS check and reconnection on network <br>errors<br> <li> Resets DNS check flag after successful login

</details>

  </td>
  <td><a href="https://github.com/TykTechnologies/tyk/pull/7136/files#diff-3b88914c99bb9418e44e6389ce73579843562e8900730b380d7fff2e95c51033">+62/-1</a>&nbsp; &nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>utils.go</strong><dd><code>Add utility functions for DNS and error handling</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

rpc/utils.go

<li>Adds utility functions for string slice comparison<br> <li> Implements network error detection logic

</details>

  </td>
  <td><a href="https://github.com/TykTechnologies/tyk/pull/7136/files#diff-659e102b74cf8cae4db311100e475bee88ba1be6dd7b450c1e060dfee2fb3761">+36/-0</a>&nbsp; &nbsp; </td>

</tr>
</table></td></tr><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>dns_resolver_test.go</strong><dd><code>Add tests for DNS resolver and reconnection logic</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

rpc/dns_resolver_test.go

<li>Adds unit tests for DNS resolution and reconnection logic<br> <li> Tests DNS change detection, error handling, and retry logic<br> <li> Covers edge cases for DNS and connection state

</details>

  </td>
  <td><a href="https://github.com/TykTechnologies/tyk/pull/7136/files#diff-1f9e4fa330ed10e933704c2e1eb7c7244fcb2e7290fab58f248fedd18f1d6f0f">+501/-0</a>&nbsp; </td>

</tr>

<tr>
  <td>
    <details>
      <summary><strong>utils_test.go</strong><dd><code>Add tests for utility functions</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

rpc/utils_test.go

<li>Adds tests for string slice comparison utility<br> <li> Adds tests for network error detection utility

</details>

  </td>
  <td><a href="https://github.com/TykTechnologies/tyk/pull/7136/files#diff-509db1da354a66f1ed878475d2ca736b989cc45f3260caf3e7933704f74f766a">+143/-0</a>&nbsp; </td>

</tr>
</table></td></tr></tr></tbody></table>

___

> <details> <summary>  Need help?</summary><li>Type <code>/help how to ...</code> in the comments thread for any questions about PR-Agent usage.</li><li>Check out the <a href="https://qodo-merge-docs.qodo.ai/usage-guide/">documentation</a> for more information.</li></details>

### CODE REVIEW

**Comment by buger** [comment]
<p>This PR is too huge for one to review :broken_heart: </p>
  
    <table>
      <tr>
          <th>Additions</th>
          <td>872 :no_good_woman: </td>
      </tr>
      <tr>
          <th>Expected</th>
          <td>:arrow_down: 800</td>
        </tr>
    </table>
    <p>
    Consider breaking it down into multiple small PRs.
    </p>
    <p>
      Check out this <a href="https://www.atlassian.com/blog/git/written-unwritten-guide-pull-requests">guide</a> to learn more about PR best-practices.
    </p>

**Comment by buger** [comment]
<p>
    Knock Knock! 🔍
  </p>
  <p>
    Just thought I'd let you know that your <em>PR title</em> and <em>story title</em> look <strong>quite different</strong>. PR titles
    that closely resemble the story title make it easier for reviewers to understand the context of the PR.
  </p>
  <blockquote>
    An easy-to-understand PR title a day makes the reviewer review away! 😛⚡️
  </blockquote>
  <table>
    <tr>
      <th>Story Title</th>
      <td>Hybrid Tyk Gateway fails to reconnect RPC when remote IP becomes unavailable</td>
    </tr>
    <tr>
        <th>PR Title</th>
        <td>TT-11285 implement dns refresh mechanism</td>
      </tr>
  </table>
  <p>
    Check out this <a href="https://www.atlassian.com/blog/git/written-unwritten-guide-pull-requests">guide</a> to learn more about PR best-practices.
  </p>

File: rpc/utils.go [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+		return false
+	}
+
+	errStr := err.Error()
+
+	// Check for specific network-related error patterns
+	return strings.Contains(errStr, "unexpected response type: <nil>. Expected *dispatcherResponse") ||
+		strings.Contains(errStr, "Cannot obtain response during timeout") ||
+		strings.Contains(errStr, "rpc is either down or was not configured") ||
+		strings.Contains(errStr, "Cannot decode response")
```
Comment (mativm02) [inline rpc/utils.go:35]: can we create variables for these and use errors.Is()?

**Review by mativm02 — APPROVED** [review]
LGTM

File: rpc/utils.go (hunk 1 above)
Comment (sredxny) [reply rpc/utils.go:35]: that would be cool,  however there is a downside with that approach, basically we need to be sure that original errors are wrapped with `fmt.Errorf(... %w ...)` and since we can't guarantee that these errors originate from our declared vars, string matching remains the most safer option here, I created this snippet to show you the issue: https://go.dev/play/p/irSblDl2NHn

**Comment by andrei-tyk** [comment]
Closed in favour of https://github.com/TykTechnologies/tyk/pull/7167 to ease the release procedure.
